lib/xandra/connection.ex

```diff
@@ -249,7 +255,7 @@ defmodule Xandra.Connection do
         {:error, {:connection_crashed, reason}}
     after
       timeout ->
-        Process.demonitor(req_alias, [:flush])
+        :gen_statem.cast(conn_pid, {:release_stream_id, stream_id})
```

Comment:
Aaah, I see okay. Couple of questions, since it's gonna take me a sec to load all of Xandra's context back into my head 😄 

1. We should still demonitor the request alias, no?
2. What happens if the response for this stream ID then comes *after the timeout*? Doesn't [this](https://github.com/whatyouhide/xandra/blob/db925408755b7be93084dd548471bf646a5eee7c/lib/xandra/connection.ex#L689-L695) happen?

Reply:
1. Yeah, the `req_alias` gets demonitored [here](https://github.com/whatyouhide/xandra/blob/8346b77c240b45cedb352bc284f0f2caaa5df77f/lib/xandra/connection.ex#L189).

Reply:
2. Good point, I thought that there was something fishy here, but didn't think it through.

- If the old stream_id that we made a request for but didn't get an answer from cassandra within `timeout` ms isn't currently used by another request, we should raise that error, yes.
- More problematic: If the old stream_id that we made a request for but didn't get an answer from cassandra within `timeout` ms is currently being used by another in_flight_request, we will be sending the answer to that new request. So we will be mixing up replies. And the fact that it gets used by another request is very likely, since we're fetching stream_ids with `Enum.at(MapSet.new(5000), 0)`, which is eventually ordered and we're using the first elements of the MapSet.

Not sure how to solve this though. Is it possible to somehow encode the req_alias to the cassandra request?

Reply:
Another idea would be to free the stream ids waaay later after the timout, so it is unlikely that we get a reply from cassandra.

Reply:
Okay, the stream_id is a short, so a reference wouldn't fit.
https://github.com/apache/cassandra/blob/trunk/doc/native_protocol_v5.spec#L318

Reply:
@harunzengin the solution here is to not free the stream ID until a response comes. See it this way: the set of usable stream IDs can be a superset of the stream IDs that we can be using *concurrently*. For example:

  * The protocol [says we have 32768 available stream IDs](https://github.com/apache/cassandra/blob/6ffd2677ee0d01c9e5493703a67255b0148106a4/doc/native_protocol_v5.spec#L331), so we can say `possible_ids = MapSet.new(1..32768)`.
  * We establish for now a limit of 5k *concurrent* requests, so we say `in_flight_ids = BoundedMapSet.new([], max_size: 5000)`, pretending we have a bounded mapset implementation.
  * We also go with a `timed_out_ids = MapSet.new([])`.
  * When a caller wants to make a req, we get an ID from `possible_ids`, at random.
  * If a req times out, we move its ID from `in_flight_ids` to `timed_out_ids` so it doesn't get picked up again. Caveat: the concurrency limit must consider `timed_out_ids` too since those reqs are in flight from the C* perspective.
  * If the response to a *timed-out request* comes later, we throw it away but move the stream ID back from `timed_out_ids` into `in_flight_ids`, because now it's actually free.

How does this sound?

Reply:
Well, this makes the assumption that the reply for each request will eventually arrive. I don't know whether this is actually the case since in our case, with the current implementation, 5000 requests got timed out within 12 hours or so and we were left with no more stream ids. What we did was a basic insert with :local_one quorum, so we expect cassandra to reply pretty quickly. This was actually the case while we were using DBConnection in the earlier versions, but as mentioned in https://github.com/whatyouhide/xandra/issues/356 the number of timeouts we are getting has also increased for a mysterious reason.

I don't exactly know whether C* will eventually reply, or whether there is a bug in this library that loses the package. My guess is that the package is lost and we don't get a reply from C*, and we'll exhaust those 5000 `in_flight_ids` the same way we do now. So, unless we find the root cause to the problem of why we're getting so many time outs, my suggestion would be to have a secondary timeout to move stream_ids from `timed_out_ids` to `possible_ids` if we don't get a reply within 30 minutes or so. How does that sound?

Reply:
> My guess is that the package is lost

In my experience, that's a pretty wild guess, because with TCP packages don't just "get lost" and Xandra is not really doing any weird routing of packages. Also:

> I don't know whether this is actually the case

This is really important. The assumption that Cassandra *never replies* doesn't seem to be based on actual data, right? Or I might totally be missing something.

We can always re-throw the timed out IDs back in the free IDs pool after a while though, yes, but I'd like to confirm that assumption first.

Reply:
> This is really important. The assumption that Cassandra never replies doesn't seem to be based on actual data, right? Or I might totally be missing something.

Well it's based on what I have experienced with the increased amount of timeouts. 

What we can do is I can implement your proposal and test it out, see whether we get an increased set of `timed_out_ids`. If they keep increasing, it's evidence that some packages get lost. If that's really the case, we can free the timed out ids after a while.

Reply:
Yep that sounds good. I think my proposal is alright anyway, then it's just a matter of whether we "expire" the timed-out IDs or figure something else out 🙃

Reply:
So, I deployed https://github.com/whatyouhide/xandra/pull/355/commits/d619ad8b52444d0e5ad38cc28f7e31dfb0a0a638 made multiple experiments but here is one of them:

```
iex8> IO.inspect(DateTime.utc_now); {:connected, state} = :sys.get_state(pid(0, 2211, 0)); nil
~U[2024-02-20 16:11:22.857108Z]
nil
iex9> t0 = state.timed_out_ids
MapSet.new([7784, 8591, 4045, 6785, 22655, 31504, 17461, 12206, 27493, 3875,
 13940, 29020, 11501, 18415, 23395, 20151, 25348, 765, 18766, 13533, 15084,
 8590, 32456, 25783, 26595, 14287, 6074, 7491, 27524, 23128, 17929, 6799, 21131,
 25779, 3533, 6170, 29558, 16523, 10859, 7700, 27722, 3818, 18573, 27986, 2673,
 9399, 26019, 11416, 14374, 31970, ...])
iex10> MapSet.size(t0)
70
iex11> IO.inspect(DateTime.utc_now); {:connected, state} = :sys.get_state(pid(0, 2211, 0)); nil
~U[2024-02-20 16:17:40.980339Z]
nil
iex12> t1 = state.timed_out_ids
MapSet.new([7784, 8591, 31802, 4045, 30009, 6785, 23810, 7058, 21662, 22655,
 31504, 17461, 27796, 12206, 27493, 27298, 3875, 27062, 13198, 8927, 13940,
 29020, 11501, 20712, 17538, 18415, 23395, 20151, 1519, 25348, 3836, 8425, 3663,
 765, 18766, 13533, 15084, 8590, 32456, 25783, 710, 26595, 14287, 11223, 6074,
 7491, 27524, 23128, 16772, 17929, ...])
iex13> MapSet.size(t1)
106
iex14> MapSet.difference(t0, t1)
MapSet.new([])
iex15> MapSet.difference(t1, t0) |> MapSet.size
36
```

So within 5 minutes, no timed out stream id has answered from cassandra. I also tried to log it in case that happens, but it got never logged over multiple experiments. My guess is that with the async way, cassandra just doesn't reply sometimes under high load, which we could address with RetryStrategy, but we need to handle the timed out stream ids and reuse them.

Reply:
@harunzengin what's the throughput you're testing at? Because ~30 just **unanswered** responses in 5 minutes is a huge number, I’m pretty confused. Also, you could try asking whether this is a possibility in the C* community, they have [a ton of ways](https://cassandra.apache.org/_/community.html) to get in touch and people are really nice.

Reply:
The rate is 1000 insertions per second. We have a cluster of 3 nodes, with 10 connections on each pool. So it's around 33 insertions per `Xandra.Connection` per second.

I have to emphasize that with DBConnection, we had 0 timeouts on our staging system, for the same 1000x insertions per second. That's the reason why I suspect the async protocol being the cause. 

I'll ask in the Cassandra community, but in the meantime, can we agree on setting a second timeout to the `timed_out_ids`, let's say 30 minutes? I already implemented it and it is ready for review.

Reply:
Question in Stackoverflow: https://stackoverflow.com/questions/78035081/concurrent-cassandra-async-writes-leading-for-some-packages-to-get-lost

Reply:
@harunzengin 30 minutes is too long. If you won't get a resp within something like 5 min, I don't see how you would (for a *single query* at least).

Reply:
@whatyouhide Reduced it to 5 minutes